### PR TITLE
Add translator from internal data structure to Opencensus format

### PR DESCRIPTION
**Pull request by dmitryax**
This implements translation from internal data representation to OpenCensus format.

Needed as a temporary solution that allows new-styles receivers to send data to old-style processors.

**Testing:** Unit tests

### CODE REVIEW

File: internal/data/trace.go [hunk 1]
```diff
@@ -135,10 +139,18 @@ func (m *Span) TraceID() TraceID {
 	return TraceIDFromBytes(m.orig.TraceId)
 }
 
+func (m *Span) TraceIDBytes() []byte {
+	return m.orig.TraceId
+}
+
```
Comment (tigrannajaryan) [inline internal/data/trace.go]: Why do we need this? What prevents us from using `TraceID()`? We will just need to add a method to get []byte from TraceID type. I think it will be nicer and will allow to keep `Span`'s exported interface minimal.

(Side note: the reason for TraceID existence is to have strong-typing to avoid confusing different ids and e.g. assigning traceid to spanid).

File: internal/data/trace.go [hunk 2]
```diff
@@ -135,10 +139,18 @@ func (m *Span) TraceID() TraceID {
...
 func (m *Span) SpanID() SpanID {
 	return SpanIDFromBytes(m.orig.SpanId)
 }
 
+func (m *Span) SpanIDBytes() []byte {
+	return m.orig.SpanId
+}
+
```
Comment (tigrannajaryan) [inline internal/data/trace.go]: Same for this.

File: internal/data/trace.go [hunk 3]
```diff
@@ -252,6 +268,14 @@ type SpanStatus struct {
 	orig *otlptrace.Status
 }
 
+func (s *SpanStatus) Code() StatusCode {
+	return StatusCode(s.orig.Code)
+}
+
+func (s *SpanStatus) Message() string {
+	return s.orig.Message
+}
```
Comment (tigrannajaryan) [inline internal/data/trace.go]: Thanks, I forgot these getters.

File: internal/data/trace.go [hunk 4]
```diff
@@ -360,10 +384,18 @@ func (m *SpanLink) TraceID() TraceID {
 	return TraceIDFromBytes(m.orig.TraceId)
 }
 
+func (m *SpanLink) TraceIDBytes() []byte {
+	return m.orig.TraceId
+}
```
Comment (tigrannajaryan) [inline internal/data/trace.go]: Same here.

File: translator/conventions/opencensus.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+const (
+	OCAttributeProcessStartTime  = "opencensus.starttime"
+	OCAttributeProcessID         = "opencensus.pid"
+	OCAttributeExporterVersion   = "opencensus.exporterversion"
+	OCAttributeResourceType      = "opencensus.resourcetype"
+	OCTimeEventMessageEventType  = "opencensus.timeevent.messageevent.type"
+	OCTimeEventMessageEventID    = "opencensus.timeevent.messageevent.id"
+	OCTimeEventMessageEventUSize = "opencensus.timeevent.messageevent.usize"
+	OCTimeEventMessageEventCSize = "opencensus.timeevent.messageevent.csize"
+)
```
Comment (tigrannajaryan) [inline translator/conventions/opencensus.go]: Looks like this is also added in my PR.

File: translator/internal/internal_to_oc.go [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+
+const sourceFormat = "ot_internal_trace"
+
+var (
+	defaultTime      = time.Time{} // zero
+	defaultTimestamp = timestamp.Timestamp{}
+	defaultProcessID = 0
+)
+
+func internalToOC(td data.TraceData) consumerdata.TraceData {
```
Comment (tigrannajaryan) [inline translator/internal/internal_to_oc.go]: We will have to return []consumerdata.TraceData here. The OC format is not capable of representing multiple batches of spans from different resources like internal data.TraceData. So the only approach that guarantees we don't lose data is to return a slice of `consumerdata.TraceData`.

File: translator/internal/internal_to_oc.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+	resourceSpansList := td.ResourceSpans()
+
+	if len(resourceSpansList) == 0 {
+		return ocTraceData
+	}
+
+	// Original resource with its spans stored in the first object of resourceSpansList
+	// Additional spans originated from different resources stored in the following objects of resourceSpansList
+	originalResourceSpans := resourceSpansList[0]
+	additionalResourceSpansList := resourceSpansList[1:]
```
Comment (tigrannajaryan) [inline translator/internal/internal_to_oc.go]: If we return a slice we will no longer need to special-case the first ResourceSpans and the rest. We can handle them uniformly.

File: translator/internal/internal_to_oc.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+			ocResource.Type = val
+		case conventions.AttributeServiceName:
+			if ocNode.ServiceInfo == nil {
+				ocNode.ServiceInfo = &occommon.ServiceInfo{}
+			}
+			ocNode.ServiceInfo.Name = val
+		case conventions.OCAttributeProcessStartTime:
+			t, err := time.Parse(time.RFC3339Nano, val)
+			if err != nil {
+				t = defaultTime
```
Comment (tigrannajaryan) [inline translator/internal/internal_to_oc.go]: I think it is probably better to leave the `StartTimestamp` field empty when it is not parseable rather than try to set it to a value.

File: translator/internal/internal_to_oc.go [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+func eventToOC(event *data.SpanEvent) *octrace.Span_TimeEvent {
+	attrs := event.Attributes()
+
+	ocMessageEventType, ocMessageEventTypeAttrFound := attrs[conventions.OCTimeEventMessageEventType]
+	ocMessageEventID, ocMessageEventIDAttrFound := attrs[conventions.OCTimeEventMessageEventID]
+	ocMessageEventUSize, ocMessageEventUSizeAttrFound := attrs[conventions.OCTimeEventMessageEventUSize]
+	ocMessageEventCSize, ocMessageEventCSizeAttrFound := attrs[conventions.OCTimeEventMessageEventCSize]
+
+	// Consider TimeEvent to be of MessageEvent type if all relevant attributes are set
```
Comment (tigrannajaryan) [inline translator/internal/internal_to_oc.go]: 👍 

Can we also check that no other attributes are set? If we don't then the attributes will be lost. If there are other attributes it is preferable to use Annotation to avoid data loss.

File: internal/internal.go [hunk 10]
```diff
@@ -38,3 +40,11 @@ func TimestampToTime(ts *timestamp.Timestamp) (t time.Time) {
 	}
 	return time.Unix(ts.Seconds, int64(ts.Nanos))
 }
+
+func TimestampToUnixnano(ts *timestamp.Timestamp) data.TimestampUnixNano {
+	return data.TimestampUnixNano(uint64(TimestampToTime(ts).UnixNano()))
+}
+
+func UnixnanoToTimestamp(u uint64) *timestamp.Timestamp {
```
Comment (tigrannajaryan) [inline internal/internal.go]: Would it make sense for this to accept data.TimestampUnixNano instead of uint64 for symmetry?

File: translator/internal/internal_to_oc.go [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+			Tracestate: traceStateToOC(link.TraceState()),
+			Attributes: attributesMapToOCSpanAttributes(link.Attributes(), link.DroppedAttributesCount()),
+		}
+		ocLinks = append(ocLinks, ocLink)
+	}
+	return ocLinks
+}
+
+func statusToOC(status data.SpanStatus) *octrace.Status {
+	emptyStatus := data.SpanStatus{}
```
Comment (tigrannajaryan) [inline translator/internal/internal_to_oc.go]: Can we move this outside the function to ensure it is not created every time (I don't know how well Go's compiler optimizes code like this).

File: internal/data/trace.go (hunk 1 above)
Comment (dmitryax) [reply internal/data/trace.go]: I wasn't sure about this. Yes, it's not that clean interface but helps to avoid immediate  wrapping->unwrapping for the usecase when only byte array is needed.

File: translator/conventions/opencensus.go (hunk 5 above)
Comment (dmitryax) [reply translator/conventions/opencensus.go]: Yes, it'll gone once it's rebased on your commit.

File: internal/data/trace.go (hunk 1 above)
Comment (tigrannajaryan) [reply internal/data/trace.go]: Can you please check if it actually does wrapping/unwrapping in the generated code or it is noop and is inlined? If it is not a noop I would get rid of it and use `[]byte` directly?

File: translator/internal/internal_to_oc.go [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+const sourceFormat = "ot_internal_trace"
```
Comment (nilebox) [inline translator/internal/internal_to_oc.go]: nit: OpenTelemetry seems shortened as `otel`, so I suggest to either change it to `otel_internal_trace`, or just `internal_trace`

File: translator/internal/internal_to_oc.go (hunk 8 above)
Comment (nilebox) [reply translator/internal/internal_to_oc.go]: @tigrannajaryan doesn't this contradict you previous suggestion for OTLP translator https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r386160820 ?
I believe @dmitryax has adapted this code from there.
We should agree on a consistent behavior and apply it everywhere.

File: translator/internal/internal_to_oc.go (hunk 6 above)
Comment (dmitryax) [reply translator/internal/internal_to_oc.go]: Makes sense if `data.TraceData` can have multiple batches of spans that don't relate to each other. I was thinking that `data.TraceData` supposed to have the same set of spans as `consumerdata.TraceData`. 

I changed it to to return slice.

File: translator/internal/internal_to_oc.go (hunk 11 above)
Comment (dmitryax) [reply translator/internal/internal_to_oc.go]: done

File: translator/internal/internal_to_oc.go (hunk 8 above)
Comment (dmitryax) [reply translator/internal/internal_to_oc.go]: Makes sense

Comment (nilebox) [reply translator/internal/internal_to_oc.go]: @tigrannajaryan doesn't this contradict you previous suggestion for OTLP translator https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r386160820 ?
I believe @dmitryax has adapted this code from there.
We should agree on a consistent behavior and apply it everywhere.

File: translator/internal/internal_to_oc.go [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+		case conventions.OCAttributeResourceType:
+			ocResource.Type = val
+		case conventions.AttributeServiceName:
+			if ocNode.ServiceInfo == nil {
+				ocNode.ServiceInfo = &occommon.ServiceInfo{}
+			}
+			ocNode.ServiceInfo.Name = val
+		case conventions.OCAttributeProcessStartTime:
+			t, err := time.Parse(time.RFC3339Nano, val)
+			if err != nil {
```
Comment (nilebox) [inline translator/internal/internal_to_oc.go]: You probably meant `if err == nil` here and below?
Also see https://github.com/open-telemetry/opentelemetry-collector/pull/614#discussion_r390688179, I suggest first discuss what is the preferred error handling and apply it both here and in OTLP translator.

Comment (nilebox) [reply translator/internal/internal_to_oc.go]: Also, it's better to use indentation for handling special cases like errors rather than for "happy path", i.e.
```
if err != nil {
  continue
}
...
``` 
It makes the code more readable and helps with avoiding typos like this one.

Comment (nilebox) [reply translator/internal/internal_to_oc.go]: See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: internal/data/trace.go (hunk 1 above)
Comment (dmitryax) [reply internal/data/trace.go]: Checked. wrapping/unwrapping appeared to be noop indeed. I changed the code accordingly

File: translator/internal/internal_to_oc.go (hunk 13 above)
Comment (dmitryax) [reply translator/internal/internal_to_oc.go]: Makes sense, thanks for catching it!

File: translator/internal/internal_to_oc.go (hunk 8 above)
Comment (dmitryax) [reply translator/internal/internal_to_oc.go]: Sure, let's agree on a consistent behavior. 

And another difference is reproducing of octrace.Span_TimeEvent_MessageEvent from span attributes. Not sure if it's needed fo OTLP->OC translator as well.

Comment (nilebox) [reply translator/internal/internal_to_oc.go]: @dmitryax See https://github.com/open-telemetry/opentelemetry-collector/pull/582#discussion_r385991303 re: MessageEvents. They are being passed as annotations as suggested there.
Is it different for internal -> OC? If so, why?

Comment (tigrannajaryan) [reply translator/internal/internal_to_oc.go]: @nilebox my earlier suggestion that you refer to asks to use a special value that indicates that it is an invalid time. This is not doing that. `defaultTime` is simply zero time, it does not indicate an invalid time. If we can indicate invalid time here it would be best. The problem is we can't, there are no invalid values for time possible. We used zero value in #582, but it did not achieve the goal that I had in my mind. My mistake that I did not catch it in the review of #582.

My suggestion in this review was to simply drop the StartTimestamp field completely instead of setting it to zero time. However, the difference between a missing field and zero time is negligible (missing field is slightly better since it saves space). Since we already have zero time implemented in #582, I am fine with doing the same here for consistency purposes. Works either way (either way is not ideal, but we can't do much if the data format does not allow us to indicate an error).

TLDR; ignore my rant and keep `defaultTime` as is.

Comment (tigrannajaryan) [reply translator/internal/internal_to_oc.go]: As for MessageEvents. I like the approach in this PR better. It attempts to do preserve the semantics as much as possible.

It may not matter much, but I think this is an improvement over hard-coding to Annotation only. We may want to reproduce this logic in OTLP->OC implementation or alternatively just live with it for now, since that code is temporary.

File: internal/data/trace.go (hunk 1 above)
Comment (tigrannajaryan) [reply internal/data/trace.go]: Good, I was worried that Go compiler will not optimize it out :-)

**Comment by tigrannajaryan** [comment]
@dmitryax please resolve the conflicts.

File: internal/data/trace.go [hunk 14]
```diff
@@ -77,12 +81,20 @@ type TraceID struct {
 	bytes []byte
 }
 
+func (t TraceID) Bytes() []byte {
+	return t.bytes
+}
+
 func TraceIDFromBytes(bytes []byte) TraceID { return TraceID{bytes} }
```
Comment (tigrannajaryan) [inline internal/data/trace.go]: Should we also rename this to NewTraceID to make it more idiomatic?

Comment (dmitryax) [reply internal/data/trace.go]: Yes, makes sense. changed

File: translator/internal/internal_to_oc.go [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+
+	// Consider TimeEvent to be of MessageEvent type if all and only relevant attributes are set
+	ocMessageEventAttrs := []string{
+		conventions.OCTimeEventMessageEventType,
+		conventions.OCTimeEventMessageEventID,
+		conventions.OCTimeEventMessageEventUSize,
+		conventions.OCTimeEventMessageEventCSize,
+	}
+	if len(attrs) == len(ocMessageEventAttrs) {
+		ocMessageEventAttrValues := map[string]data.AttributeValue{}
```
Comment (pjanotti) [inline translator/internal/internal_to_oc.go]: since the keys come in a fixed order you could do with an array instead, but probably not worth the trouble (better do some benchmark before changing it)
